[GStreamer] Try to detect and recover inconsistent playing state

calvaris opened the pull request:
#### c5cfffe19659bb95742d4d5ebc46c405d329fbed
<pre>
[GStreamer] Try to detect and recover inconsistent playing state
<a href="https://bugs.webkit.org/show_bug.cgi?id=252310">https://bugs.webkit.org/show_bug.cgi?id=252310</a>

Reviewed by Philippe Normand.

A series of buffer flushings could leave pipeline in an inconsistent state: top-level bin in playing with sinks in
paused. This happens if flush is triggered in the middle of paused-to-playing transition which was started by gst bin
async handling of ASYNC_DONE message.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::areAllSinksPlayingForBin):
(WebCore::MediaPlayerPrivateGStreamer::checkPlayingConsistency):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/261571@main">https://commits.webkit.org/261571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce81f17b19f3ec7d99d92c8ed877ab53b36d8474

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107933 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40817 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117066 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116413 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111820 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18377 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8308 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100121 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95783 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28733 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41690 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9914 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30081 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83441 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10621 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6970 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49670 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8066 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12193 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4160,6 +4161,63 @@ std::optional<VideoFrameMetadata> MediaPlayerPrivateGStreamer::videoFrameMetadat
     return metadata;
 }
 
+void MediaPlayerPrivateGStreamer::checkPlayingConsitency()
```
Use `GstIteratorAdaptor` here please :)

calvaris replied:
Nice request. Code is indeed much smaller and easier to understand.

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1808,6 +1808,7 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
         if (!m_isLegacyPlaybin && currentState == GST_STATE_PAUSED && newState == GST_STATE_PLAYING)
             playbin3SendSelectStreamsIfAppropriate();
         updateStates();
+        checkPlayingConsitency();
```
The bug title has MSE tag so maybe do this only for `isMediaSource()`?

Also, there is a typo in the method name.

calvaris replied:
Nice typo catch. We could do it for `isMediaSource()` only but I think it's a good sanity check even for regular playback. I would change the bug title instead, wdyt?

philn replied:
Whatever... as long as this doesn't impact perf...

calvaris committed:
[GStreamer] Try to detect and recover inconsistent playing state
https://bugs.webkit.org/show_bug.cgi?id=252310

Reviewed by Philippe Normand.

A series of buffer flushings could leave pipeline in an inconsistent state: top-level bin in playing with sinks in
paused. This happens if flush is triggered in the middle of paused-to-playing transition which was started by gst bin
async handling of ASYNC_DONE message.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::areAllSinksPlayingForBin):
(WebCore::MediaPlayerPrivateGStreamer::checkPlayingConsistency):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/261571@main
